File: core/src/main/java/com/emarsys/core/DefaultCoreCompletionHandler.java
package com.emarsys.core;

import com.emarsys.core.api.ResponseErrorException;
import com.emarsys.core.api.result.CompletionListener;
import com.emarsys.core.request.model.RequestModel;
import com.emarsys.core.response.ResponseModel;
import com.emarsys.core.util.Assert;

import java.util.Map;

public class DefaultCoreCompletionHandler implements CoreCompletionHandler, Registry<RequestModel, CompletionListener> {
    private final Map<String, CompletionListener> completionListenerMap;

    public DefaultCoreCompletionHandler(Map<String, CompletionListener> completionListenerMap) {
        Assert.notNull(completionListenerMap, "CompletionListenerMap must not be null!");
        this.completionListenerMap = completionListenerMap;
    }

    public void register(RequestModel model, CompletionListener listener) {
        Assert.notNull(model, "RequestModel must not be null!");
        if (listener != null) {
            this.completionListenerMap.put(model.getId(), listener);
        }
    }

    @Override
    public void onSuccess(final String id, final ResponseModel responseModel) {
        callCompletionListener(id, null);
    }

    @Override
    public void onError(final String id, final Exception cause) {
        callCompletionListener(id, cause);
    }

    @Override
    public void onError(final String id, final ResponseModel responseModel) {
        Exception exception = new ResponseErrorException(
                responseModel.getStatusCode(),
                responseModel.getMessage(),
                responseModel.getBody());

        callCompletionListener(id, exception);
    }

    private void callCompletionListener(String id, Exception cause) {
        final CompletionListener listener = completionListenerMap.get(id);
        if (listener != null) {
            listener.onCompleted(cause);
            completionListenerMap.remove(id);
        }
    }

}

File: core/src/main/java/com/emarsys/core/DefaultCoreCompletionHandler.kt
package com.emarsys.core

import com.emarsys.core.api.ResponseErrorException
import com.emarsys.core.api.result.CompletionListener
import com.emarsys.core.request.model.RequestModel
import com.emarsys.core.response.ResponseModel
import java.lang.Exception

open class DefaultCoreCompletionHandler(
        private val completionListenerMap: MutableMap<String, CompletionListener>
) : CoreCompletionHandler, Registry<RequestModel, CompletionListener?> {
    override fun register(model: RequestModel, listener: CompletionListener?) {
        if (listener != null) {
            completionListenerMap[model.id] = listener
        }
    }

    override fun onSuccess(id: String, responseModel: ResponseModel) {
        callCompletionListener(id, null)
    }

    override fun onError(id: String, cause: Exception) {
        callCompletionListener(id, cause)
    }

    override fun onError(id: String, responseModel: ResponseModel) {
        val exception: Exception = ResponseErrorException(
                responseModel.statusCode,
                responseModel.message,
                responseModel.body)
        callCompletionListener(id, exception)
    }

    private fun callCompletionListener(id: String, cause: Exception?) {
        val listener = completionListenerMap[id]
        if (listener != null) {
            listener.onCompleted(cause)
            completionListenerMap.remove(id)
        }
    }
}